Add infrastructure for trimming and NativeAOT test apps.

eerhardt opened the pull request:
This infrastructure was taken from https://github.com/dotnet/runtime/tree/c62f69be1405a8e41b56ffc05f22d791bf4c7d2d/eng/testing/linker and modified to work in dotnet/aspnetcore.

Added the first test:
- Generic host + value type container builder (verify that error is thrown)

Contributes to #45860

It is probably easiest to compare the files in `eng/testing/linker` to the files in dotnet/runtime to see what I changed. I can prepare that in a repo if people want. Or just use Beyond Compare (or similar) to diff between the two repos.

Working on hooking this into CI now.

eerhardt committed:
Add infrastructure for trimming and NativeAOT test apps.

This infrastructure was taken from https://github.com/dotnet/runtime/tree/c62f69be1405a8e41b56ffc05f22d791bf4c7d2d/eng/testing/linker and modified to work in dotnet/aspnetcore.

Added the first test:
- Generic host + value type container builder (verify that error is thrown)

Contributes to #45860

joperezr commented on eng/testing/linker/SupportFiles/Directory.Build.targets:
```diff
@@ -0,0 +1,25 @@
+  This allows for easy dev workflow. Just need to update the product code, and re-run the tests.
+  -->
+  <ItemGroup>
+    <ProjectReference Include="$(RepoRoot)src\DefaultBuilder\src\Microsoft.AspNetCore.csproj" />
```
Don't we also need to remove the existing FrameworkReference for this to work? Or does conflict resolution automatically picks the greater version here?

eerhardt replied:
So I waffled back and forth on how I wanted to do this. High-level there are 2 options:

1. Reference the ASP.NET assemblies through the "shared framework" that is produced by the full build.
    - This is how customers make ASP.NET apps.
    - However this makes the dev workflow less than ideal. ASP.NET doesn't have the "bin place" logic that dotnet/runtime has. So in order to change code and re-run the tests, you would need to do a full build, or copy files to the right spot.
2. Don't be a "Web" SDK project, just a plain Microsoft.NET.Sdk and ProjectReference each ASP.NET assembly
    - This is how our existing tests and samples work

I choose to go with (2) because of the simplicity and because of the better dev workflow.

joperezr replied:
Makes sense. For pure education purposes, do you know what else is different apart from reference assemblies when not being a web sdk project? For instance, does it change how dotnet run works?

eerhardt replied:
One example of things that are different is that some MSBuild properties get defaulted differently. For example "Server GC" gets enabled. Currently the "disable Reflection in JsonSerializer" feature switch gets defaulted in the Web SDK as well.

For these cases, I'm not too concerned. For the feature switches, we can default them as needed in the tests, or in the infrastructure.

joperezr commented on eng/testing/linker/SupportFiles/Directory.Build.targets:
```diff
@@ -0,0 +1,25 @@
+    <SuppressGenerateILCompilerExplicitPackageReferenceWarning>true</SuppressGenerateILCompilerExplicitPackageReferenceWarning>
+  </PropertyGroup>
+
+  <!-- needed to reference a specific version of NetCoreApp. Workaround https://github.com/dotnet/runtime/issues/81382 -->
```
According to a @sbomer comment on this issue this will be fixed (or has already been fixed) soon. Is the expectation that once it is we can remove this whole thing and just add a RuntimeFrameworkVersion property to the project.csproj template?

eerhardt replied:
That is the hope. However this is blocked because the ASP.NET and Windows.Desktop try to get restored with the same version, which they don't have matching versions on daily builds. So we need a better story here how to reference a daily build of dotnet/runtime. That's the hope that will be fixed with the referenced issue.

joperezr commented on src/DefaultBuilder/test/Microsoft.AspNetCore.NativeAotTests/Microsoft.AspNetCore.NativeAotTests.proj:
```diff
@@ -0,0 +1,9 @@
+<Project DefaultTargets="Build">
+  <Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Build.props))" />
```
I assume this is required because of the fact that this is not an SDK project. I know you are just copying what we do in dotnet/runtime, but I wonder if instead we should just use the NoTargets SDK instead that way these imports are automatic.

eerhardt replied:
Check out my latest commit. I tried using NoTargets and Traversal, but ended up being a normal Microsoft.NET.Sdk project. I needed to do something different in order to get the `ExcludeFromBuild` arcade logic to work correctly. The way aspnetcore's build works is a lot different than how dotnet/runtime's build works. runtime is more "top-down" and aspnetcore is more "bottom-up".

joperezr commented on src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:
```diff
@@ -86,7 +86,7 @@ public static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Action
     /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
     /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-    /// <remarks>When using the il linker, all public methods of <typeparamref name="TStartup"/> are preserved. This should match the Startup type directly (and not a base type).</remarks>
+    /// <remarks>When in a trimmed app, all public methods of <typeparamref name="TStartup"/> are preserved. This should match the Startup type directly (and not a base type).</remarks>
```
I'm not very familiar how these comments work for the aspnetcore repo in terms of what shows up in intellisense. Are triple slash comments the source of truth? If not, and if instead we use a model similar to runtime where we have api-docs being the source of truth, then perhaps we should make sure that these updates in comments get applied there too.

eerhardt replied:
Hmm, that's a good question. I assumed that the triple-slash comments were the source of truth because I've never seen anyone updating them anywhere else. Anyone else know?

JamesNK replied:
XML comments are the source of truth in dotnet/aspnetcore.

joperezr commented on eng/testing/linker/trimmingTests.targets:
```diff
@@ -0,0 +1,133 @@
+    <Message Text="Generated $(_projectFile)" />
+  </Target>
+
+  <Target Name="GetTrimmingProjectsToRestore"
```
Can you remind me where is this target used?

eerhardt replied:
Looks like this is more "dotnet/runtime-specific" logic. 

https://github.com/dotnet/runtime/blob/34138a15b1624c533b31658915e4f60aa7562614/src/libraries/tests.proj#L772-L780

I will remove it here.

joperezr reviewed (approved):
Left some minor comments and questions, but other than that (and assuming the builds work 😃) this LGTM.

eerhardt committed:
Skip IsPublishedAppTestProject projects when SkipTestBuild=true.

Use `Sdk="Microsoft.NET.Sdk"` and hook into CoreBuild for the published app test projects in order to skip the build correctly.

eerhardt committed:
Enforce build order

Ensure the IsPublishedAppTestProject projects aren't built until the shared fx is built.

eerhardt committed:
try depending on Microsoft.AspNetCore.csproj for build ordering

eerhardt committed:
Try using RequiresDelayedBuild

eerhardt committed:
Clean up hacks

eerhardt committed:
Add workaround for https://github.com/dotnet/arcade/pull/13406/

eerhardt committed:
Only run the AOT tests during Test.

Skip "Build" and skip pubilshing for Helix.

mitchdenny commented:
So if I am reading this correctly, we basically have a set of console apps/ASP.NET apps in a single source file that we generate a project for, then publish (with AOT) and run them. And the return code from the program when it exits indicates success or failure?

eerhardt commented:
> So if I am reading this correctly, we basically have a set of console apps/ASP.NET apps in a single source file that we generate a project for, then publish (with AOT) and run them. And the return code from the program when it exits indicates success or failure?

💯 percent correct.

eerhardt commented on Directory.Build.props:
```diff
@@ -260,6 +264,7 @@
   <Import Project="eng\targets\Java.Common.props"  Condition="'$(MSBuildProjectExtension)' == '.javaproj'" />
   <Import Project="eng\targets\Helix.props" Condition=" $(IsTestProject) " />
   <Import Project="eng\targets\FunctionalTestWithAssets.props" Condition=" $(IsTestProject) " />
+  <Import Project="eng\testing\linker\trimmingTests.props" Condition="'$(IsPublishedAppTestProject)' == 'true'" />
```
```suggestion
  <Import Project="eng\testing\linker\trimmingTests.props" Condition="'$(IsPublishedAppTestProject)' == 'true'" />
  <Import Project="eng\targets\Helix.props" Condition=" $(IsTestProject) " />
  <Import Project="eng\targets\FunctionalTestWithAssets.props" Condition=" $(IsTestProject) " />
```

To be consistent with how the .targets files are imported.

eerhardt commented on eng/testing/linker/trimmingTests.targets:
```diff
@@ -0,0 +1,132 @@
+<Project>
+  <Import Project="$(RepoRoot)eng\targets\ResolveReferences.targets" />
```
Ensure this is still needed. I think it is needed for `CodeGen.proj` to work correctly.

eerhardt committed:
PR feedback

wtgodbe commented on eng/testing/linker/SupportFiles/Directory.Build.targets:
```diff
@@ -0,0 +1,25 @@
+  Reference the Microsoft.AspNetCore.App shared framework assemblies with ProjectReference.
+  This allows for easy dev workflow. Just need to update the product code, and re-run the tests.
+  -->
+  <ItemGroup>
```
Did app.ref not work?

eerhardt replied:
It works, but it is a bit heavier than this approach.

For example, the "incremental" run of `.\.dotnet\dotnet.exe test .\src\DefaultBuilder\test\Microsoft.AspNetCore.NativeAotTests\` takes ~15s using app.ref and takes ~10s with this approach.

We can always switch this as we go forward if we need to. But for now I'd rather do the minimal.

wtgodbe reviewed (approved):
Seems reasonable enough to me
